### DISCUSSION TITLE
Python API fetchall() consumes too much time

### DISCUSSION THREAD

**Discussion by amoghnjoshi** [duckdb/duckdb#7737] [Q&A]
I am using the python API in below fashion.
Dataset: Around 40 million rows with 20 columns each.
Using single python connection (read_only mode) with 8/16 threads.

          conn.execute(query)
          column_names = conn.description
            for query_result in conn.fetchall():
                for count, item in enumerate(query_result):
                    // process the individual query_result

It is observed that for some of the queries, the fetchall() takes much time (even more than that taken by execute()). 
Interestingly, there are no (zero) results returned by the conn.fetchall() method and that is correct as per the dataset and the query used. The queries involve use of group by and aggregate functions.

Any hints on why the conn.fetchall() might be consuming so much time and how to avoid the same?

**Comment by Tishj** [comment]
What types of queries are causing this?
are these SELECTS or what are the statement types?

I'm assuming this issue is similar to #7566, where `execute` executes the query and there is no result, so it doesn't populate the internal result.
Then `fetchall` tries to find the result and re-executes if it cant find it

**Reply by amoghnjoshi** [reply]
Yes, they are SELECT queries with some filters and group by.
Is there any workaround to avoid such double query execution?

**Reply by Tishj** [reply]
It sounds like a bug, but without a reproduction example I can't be sure

**Reply by Tishj** [reply]
In the meantime, try creating a relation instead, and fetching from that.
If the problem is what I think, that should fix it.

```py
rel = conn.sql(query)
for query_result in rel.fetchall(): ...
```
